File: client/client-multi/client-java/src/main/java/org/kaaproject/kaa/client/Kaa.java
/*
 * Copyright 2014 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client;

import java.io.IOException;
import java.util.Properties;

/**
 * Entry point to the Kaa library.
 *
 * Responsible for the Kaa initialization and start/stop actions.
 *
 * @author Yaroslav Zeygerman
 *
 */
public class Kaa {
    private static DefaultKaaClient client;

    /**
     * Initialize Kaa library using given properties
     *
     * @param properties properties for the initialization
     * @throws Exception
     *
     */
    public static void init(Properties properties) throws Exception {
        if (client != null) {
            client.stop();
        }
        client = new DefaultKaaClient(properties);
        client.init();
    }

    /**
     * Starts Kaa's work flow.
     *
     */
    public static void start() throws IOException {
        if (client != null) {
            client.start();
        }
    }

    /**
     * Stops Kaa's work flow.
     *
     */
    public static void stop() {
        if (client != null) {
            client.stop();
        }
    }

    /**
     * Retrieves the Kaa client.
     *
     * @return Kaa client.
     *
     */
    public static KaaClient getClient() {
        return client;
    }
}

File: client/client-multi/client-cpp/impl/Kaa.cpp
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "kaa/Kaa.hpp"

#include <memory>

#include "kaa/KaaClient.hpp"
#include "kaa/common/exception/KaaException.hpp"
#include "kaa/KaaClientPlatformContext.hpp"

namespace kaa {

Botan::LibraryInitializer Kaa::botanInit_("thread_safe=true");

std::shared_ptr<IKaaClient> Kaa::newClient(IKaaClientPlatformContextPtr context
                                         , KaaClientStateListenerPtr listener)
{
    if (!context) {
        throw KaaException("Kaa client platform context is null");
    }
    return std::shared_ptr<IKaaClient>(new KaaClient(context, listener));
}

}
